opm/material/densead/DynamicEvaluation.hpp

```diff
@@ -0,0 +1,479 @@
+
+    // create a "blank" dynamic evaluation
+    explicit Evaluation(int numDerivatives)
+        : data_(1 + numDerivatives)
+    {}
+
+    // create a dynamic evaluation which represents a constant function
+    //
+    // i.e., f(x) = c. this implies an evaluation with the given value and all
+    // derivatives being zero.
+    template <class RhsValueType>
+    Evaluation(int numDerivatives, const RhsValueType& c)
+        : data_(1 + numDerivatives, 0.0)
+    {
+        //clearDerivatives();
+        setValue(c);
+
+        checkDefined_();
+    }
+
+    // create an evaluation which represents a constant function
+    //
+    // i.e., f(x) = c. this implies an evaluation with the given value and all
+    // derivatives being zero.
+    template <class RhsValueType>
+    Evaluation(int numVars, const RhsValueType& c, int varPos)
+     : data_(1 + numVars, 0.0)
+    {
+        // The variable position must be in represented by the given variable descriptor
+        assert(0 <= varPos && varPos < size());
+
+        setValue(c);
+
+        data_[varPos + dstart_()] = 1.0;
+
+        checkDefined_();
+    }
+
+    // set all derivatives to zero
+    void clearDerivatives()
+    {
+        for (int i = dstart_(); i < dend_(); ++i)
+            data_[i] = 0.0;
+    }
+
+    // create a function evaluation for a "naked" depending variable (i.e., f(x) = x)
+    template <class RhsValueType>
+    static Evaluation createVariable(int numVars, const RhsValueType& value, int varPos)
+    {
+        // copy function value and set all derivatives to 0, except for the variable
+        // which is represented by the value (which is set to 1.0)
+        return Evaluation(numVars, value, varPos);
+    }
+
+
+    // "evaluate" a constant function (i.e. a function that does not depend on the set of
+    // relevant variables, f(x) = c).
+    template <class RhsValueType>
+    static Evaluation createConstant(int numVars, const RhsValueType& value)
+    {
+        return Evaluation(numVars, value);
+    }
+
+    // print the value and the derivatives of the function evaluation
+    void print(std::ostream& os = std::cout) const
+    {
+        // print value
+        os << "v: " << value() << " / d:";
+
+        // print derivatives
+        for (int varIdx = 0; varIdx < size(); ++varIdx) {
+            os << " " << derivative(varIdx);
+        }
+    }
+
+    // copy all derivatives from other
+    void copyDerivatives(const Evaluation& other)
+    {
+        for (int i = dstart_(); i < dend_(); ++i)
+            data_[i] = other.data_[i];
+    }
+
+
+    // add value and derivatives from other to this values and derivatives
+    Evaluation& operator+=(const Evaluation& other)
+    {
+        for (int i = 0; i < length_(); ++i)
+            data_[i] += other.data_[i];
+
+        return *this;
+    }
+
+    // add value from other to this values
+    template <class RhsValueType>
+    Evaluation& operator+=(const RhsValueType& other)
```

Comment:
Should we just use `ValueType` make the functions with `RhsValueType` not template function anymore?  Basically, we can only have operations with `Scalar` and `Evaluation` with the same number of primary variables.

Reply:
this has not been changed by this PR. the reason for the templated type of the right-hand side is that multiplying with e.g. integers while ValueType is e.g. double could be more efficient than forcing a cast to ValueType. (also it is conceivable that types cannot be cast into each other but that the required operators are available-.)

Reply:
> this has not been changed by this PR. the reason for the templated type of the right-hand side is that multiplying with e.g. integers while ValueType is e.g. double could be more efficient than forcing a cast to ValueType.

I am not totally sure about this.  I think casting will happen anyway for the case you mentioned. Either happens when calling the function, or happens inside the function.  For some operations like `*=`, which including the interactions with derivatives, casting might happen more than once. 

I am not sure why it will be more efficient with templated type. 

It is just some thoughts on trival levels. I can be wrong.

Reply:
It might also reduce some compilation time potentially?

Reply:
> I am not totally sure about this. I think casting will happen anyway for the case you mentioned. Either happens when calling the function, or happens inside the function. 

also consider that this code supports nested AD and that this is used in some (non-flow) places. casting an integer to an evaluation is no longer be guaranteed to work with dynamic evaluations and the result will be quite a bit slower even for static sizes.

> It might also reduce some compilation time potentially?

my hunch is that this will be in the order of a few milliseconds. having to recompile the whole library of an upstream module bothers me much more...